src/app/shared/components/autocomplete/autocomplete.component.ts

```diff
@@ -0,0 +1,103 @@
+import {
+  Component,
+  Input,
+  OnInit,
+  forwardRef,
+  OnDestroy,
+  HostBinding,
+  Injector,
+  Type
+} from '@angular/core';
+import {
+  ControlValueAccessor,
+  FormControl,
+  NG_VALUE_ACCESSOR,
+  NgControl
+} from '@angular/forms';
+import { FormFieldControl } from '@app/shared/components/control';
+import { Subject } from 'rxjs';
+import { takeUntil, tap } from 'rxjs/operators';
+
+@Component({
+  selector: 'vtm-autocomplete',
+  templateUrl: './autocomplete.component.html',
+  styleUrls: ['./autocomplete.component.scss'],
+  providers: [
+    {
+      provide: NG_VALUE_ACCESSOR,
+      useExisting: forwardRef(() => AutocompleteComponent),
+      multi: true
+    },
+    {
+      provide: FormFieldControl,
+      useExisting: AutocompleteComponent
+    }
+  ]
+})
+export class AutocompleteComponent
+  implements ControlValueAccessor, FormFieldControl, OnInit, OnDestroy {
+  @Input() autocompleteData;
+  @Input() hasError: boolean;
+  @Input('aria-describedby') ariaDescribedBy: string | null;
+
+  autocompleteInput: FormControl;
+  errorMessages?: { rule: string; message: string }[];
+  controlType = 'text';
+  multi = true;
+
+  @Input('value') _value = '';
+  set value(val) {
+    this._value = val;
+    this.onchange(val);
+  }
+  get value() {
+    return this._value;
+  }
+
+  @HostBinding('attr.id')
+  externalId = '';
+
+  @Input()
+  set id(value: string) {
+    this.internalId = value;
+    this.externalId = null;
+  }
+  get id() {
+    return `${this.internalId}`;
+  }
+  private internalId = '';
```

Comment:
can we move all the initialisation or definition at the top of the file so it is consistent and we know what to expect as a behaviour from the component please? 
For example, private variables, ngForm, @Input, @Output, getters, setters, constructor, methods.
It will make things easier to reason about if we can identify them straight away